File: 백준/단계별/동적 계획법 1/1로 만들기.cpp
#include<stdio.h>
int min(int x,int y){return x<y?x:y;}
int arr[1000001]={0,0,};
int main()
{
	int x;
	scanf("%d",&x);
	for(int i=2;i<=x;i++)
	{
		arr[i]=arr[i-1]+1;
		if(i%3==0){
			arr[i]=min(arr[i],arr[i/3]+1);
		}
		if(i%2==0){
			arr[i]=min(arr[i],arr[i/2]+1);
		}
	}
	printf("%d",arr[x]);
	return 0;
}

File: 백준/단계별/동적 계획법 1/1로 만들기.py
x = int(input())
dp = [0 for _ in range(10000000)]

for i in range(2, x+1):
    dp[i] = dp[i-1] + 1
    if i % 2 == 0:
        dp[i] = min(dp[i], dp[int(i/2)] + 1)
    if i % 3 == 0:
        dp[i] = min(dp[i], dp[int(i/3)] + 1)
print(dp[x])